### PR TITLE
gh-90656: Add platform triplets for 64-bit LoongArch (LA64)

### LINKED ISSUE #90656: Add new triplets for 64-bit LoongArch
BPO | [46498](https://bugs.python.org/issue46498)
--- | :---
Nosy | @loongson-zn, @xen0n
PRs | <li>python/cpython#30939</li>
Files | <li>[loongarch64.diff](https://bugs.python.org/file50579/loongarch64.diff "Uploaded as text/plain at 2022-01-24.05:00:21 by @loongson-zn")</li><li>[loongarch64-new.diff](https://bugs.python.org/file50590/loongarch64-new.diff "Uploaded as text/plain at 2022-01-27.08:05:29 by @loongson-zn"): update loongarch64.diff</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-01-24.05:00:22.024>
labels = ['build', '3.9']
title = 'Add new triplets for loongarch64'
updated_at = <Date 2022-01-27.08:05:30.042>
user = 'https://github.com/loongson-zn'
```

bugs.python.org fields:
```python
activity = <Date 2022-01-27.08:05:30.042>
actor = 'loongson-zn'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Build']
creation = <Date 2022-01-24.05:00:22.024>
creator = 'loongson-zn'
dependencies = []
files = ['50579', '50590']
hgrepos = []
issue_num = 46498
keywords = ['patch']
message_count = 2.0
messages = ['411701', '411842']
nosy_count = 2.0
nosy_names = ['loongson-zn', 'xen0n']
pr_nums = ['30939']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'compile error'
url = 'https://bugs.python.org/issue46498'
versions = ['Python 3.9']
```

</p></details>

### PR DESCRIPTION
**Pull request by loongson-zn**
Add new triplets for loongarch64, please review, thanks!
https://bugs.python.org/issue46498

https://bugs.python.org/issue46498

Signed-off-by: Zhang Na <zhangna@loongson.cn>
Reviewed-by: Wang Xuerui   <git@xen0n.name>
Reviewed-by: Wu Xiaotian <wuxiaotian@loongson.cn>
Reviewed-by: Zhu Yaliang   <zhuyaliang@loongson.cn>

* Issue: gh-90656

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@loongson-zn

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by loongson-zn** [comment]
Okay, I'm working on it now.

**Comment by cheungxi** [comment]
ping

File: configure.ac
Comment (xen0n) [inline configure.ac]: We recently discovered this symbol is in fact deprecated. Please use the more expressive `__loongarch_lp64` instead.

File: Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst
Comment (xen0n) [inline Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst]: This, and the PR title, could probably be "Add platform triplets for LoongArch64" -- shorter and properly capitalized. (Things just added are implicitly "new".)

**Review by xen0n — COMMENTED** [review]
Please **rebase** on top of main branch so that no merge commits remain in this PR.

File: configure.ac
Comment (loongson-zn) [reply configure.ac]: Ok! When the final version of the manual is released, we will be adjusted together. Please wait a few more days!

**Comment by loongson-zn** [comment]
[LoongArch toolchain conventions](https://loongson.github.io/LoongArch-Documentation/LoongArch-toolchain-conventions-EN.html#_gnu_target_triplets_and_multiarch_specifiers) -- official documentation instructions, please review, thanks! @corona10  @erlend-aasland

**Comment by corona10** [comment]
@loongson-zn I will take a look, but it may takes some times to review this PR :)
Please keep patient

**Comment by loongson-zn** [comment]
> @loongson-zn I will take a look, but it may takes some times to review from the view of supporting this platform is proper. Please keep patient

Thanks，this is [debian documentation](https://wiki.debian.org/Multiarch/Tuples)  about LoongArch,  we need cpython support LoongArch.

**Comment by loongson-zn** [comment]
[llvm 16.0.0 ](https://releases.llvm.org/16.0.0/docs/ReleaseNotes.html#changes-to-the-loongarch-backend)、[kernel 5.19](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.27&id=c6f2f3e2c80e975804360665d973211e4d9390cb)、[gcc 13.1](https://gcc.gnu.org/gcc-13/)、[binutils 2.40](https://ftp.gnu.org/gnu/binutils) 、[glibc 2.36](https://mail.gnu.org/archive/html/info-gnu/2022-08/msg00000.html) and many other projects already support LoongArch.
On the other hand, I find python in the [Gentoo](https://www.gentoo.org/downloads/#loong) , `. /python3.11-config --extension-suffix `output is `.cpython-311.so`, but to maintain consistency with other architectures, the output should be `.cpython-311-loongarch64-linux-gnu.so`. So, I think this commit should be merged as soon as possible to avoid more new questions continue to arise.

**Comment by loongson-zn** [comment]
rebase update

**Review by xen0n — COMMENTED** [review]
The commit message tags are malformed. Email addresses must be attached:

```
Co-authored-by: WANG Xuerui <git@xen0n.name>
Signed-off-by: Zhang Na <zhangna@loongson.cn>
```

Also I don't know if CPython uses DCO tags (i.e. `Signed-off-by`) by convention. If it's not the case you may safely drop that line. I can confirm the other part (the bulk of `#ifdef`-ery) was suggested and written by me though.

**Comment by loongson-zn** [comment]
Yeah, you are right. I will make corrections.

**Comment by arhadthedev** [comment]
> Also I don't know if CPython uses DCO tags

It doesn't.

**Comment by erlend-aasland** [comment]
Oleg, the `build` label is for issues, not PRs. For PRs, we clearly see which part of the code base is touched from the [diff view](https://github.com/python/cpython/pull/30939/files), so `build` is superfluous; it only clutters the GH issue/PR search, contributing to noise. I consider the `topic-` labels likewise[^1]. The same logic applies: we clearly see that a PR is a `topic-argument-clinic` if it touches clinic.py; the clutter argument also applies.

[^1]: though the devguide says they are for issues/PRs, they are de facto used for issues, not PRs

**Comment by arhadthedev** [comment]
> the build label is for issues, not PRs.
>
> I consider the topic- labels likewise.

Thank you, got it.

**Comment by loongson-zn** [comment]
> [llvm 16.0.0 ](https://releases.llvm.org/16.0.0/docs/ReleaseNotes.html#changes-to-the-loongarch-backend)、[kernel 5.19](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.1.27&id=c6f2f3e2c80e975804360665d973211e4d9390cb)、[gcc 13.1](https://gcc.gnu.org/gcc-13/)、[binutils 2.40](https://ftp.gnu.org/gnu/binutils) 、[glibc 2.36](https://mail.gnu.org/archive/html/info-gnu/2022-08/msg00000.html) and many other projects already support LoongArch. On the other hand, I find python in the [Gentoo](https://www.gentoo.org/downloads/#loong) , `. /python3.11-config --extension-suffix `output is `.cpython-311.so`, but to maintain consistency with other architectures, the output should be `.cpython-311-loongarch64-linux-gnu.so`. So, I think this commit should be merged as soon as possible to avoid more new questions continue to arise.

@erlend-aasland @corona10 @arhadthedev As far as I know, LoongArch is promoting community work in Debian and Yocto. Can we consider accelerating the progress of Python's support work?

**Comment by corona10** [comment]
Sorry for the late response:
Please consider that you can meet the requirement of [PEP-11](https://peps.python.org/pep-0011/)

At least for tier 3, you would need at least one core dev and reliable build bot.

**Comment by corona10** [comment]
> Thanks，this is [debian documentation](https://wiki.debian.org/Multiarch/Tuples) about LoongArch, we need cpython support LoongArch.

My suggestion is how sending a downstream patch to Debian first?

**Comment by erlend-aasland** [comment]
> Sorry for the late response: Please consider that you can meet the requirement of [PEP-11](https://peps.python.org/pep-0011/)
> 
> At least for tier 3, you would need at least one core dev and a reliable buildbot first.

PEP-11 is for platforms, not architectures?

**Comment by corona10** [comment]
> [PEP-11](https://peps.python.org/11) is for platforms, not architectures?

1. Who will maintain the loongarch64-unknown-linux-gnu ?
2. How can we verify the code for loongarch64-unknown-linux-gnu ?

I am sure that such requirements should be checked before we accept the related code from the maintenance cost view.

**Comment by corona10** [comment]
And we requested similar requirements to Windriver team too
: https://mail.python.org/pipermail/python-dev/2019-January/156031.html

**Comment by corona10** [comment]
@erlend-aasland  cc @loongson-zn 

But on the other hand, if this is the only patch that they will need, I am okay to accept the patch anyway.

**Comment by erlend-aasland** [comment]
> And we requested similar requirements to Windriver team too
> : https://mail.python.org/pipermail/python-dev/2019-January/156031.html

AFAICS, that's also about a platform, not an architecture ;) Anyways, this is getting too offtopic :)

**Comment by xen0n** [comment]
My two cents:

> Who will maintain the loongarch64-unknown-linux-gnu?

Aside from the Loongson teams, there is [an active community around LoongArch](https://bbs.loongarch.org/) (also [Telegram user group](https://t.me/loongson_users)) with multiple porters and power users willing to help. I'm heavily involved in and contributed to many upstream projects' LoongArch support, if it helps.

> How can we verify the code for loongarch64-unknown-linux-gnu?

Multiple community members regularly test out new package versions on real hardware, in addition to in-house testing by Loongson and their partners.

And with my Gentoo hat on: I regularly test Python packages as they're updated upstream. And as Gentoo heavily depends on Python, any major breakage should be noticeable once it reaches the port's userbase.

File: Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst]:
```suggestion
Add platform triplets for ``LoongArch64``:

- ``loongarch64-linux-gnusf``
- ``loongarch64-linux-gnuf32``
- ``loongarch64-linux-gnu``

Patch by Zhang Na.
```

Also add a similar entry to Doc/whatsnew/3.12.rst.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]

**Comment by erlend-aasland** [comment]
> rebase update

BTW, for the CPython repo, please don't force push; use `git merge --no-ff main` instead :) Force-pushing does not play well with the GitHub UI. Also, we squash merge all PRs into a single commit, so it does not matter if the PR commit history is cluttered. See also [the devguide](https://devguide.python.org).

File: Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst]: The What's New category should be "Build Changes", IMO.

**Comment by xen0n** [comment]
> > Aside from the Loongson teams, there is [an active community around LoongArch](https://bbs.loongarch.org/) (also [Telegram user group](https://t.me/loongson_users)) with multiple porters and power users willing to help. I'm heavily involved in and contributed to many upstream projects' LoongArch support, if it helps.
> 
> Yeah, I understand how you try to support the LoongArch for various projects. But maintaining from the CPython is a different story, so what I want to confirm about accepting the patch is that all you have to do is adding a triplet, or do you need more patches in the future that include modifying CPython code with [preprocessor](https://loongson.github.io/LoongArch-Documentation/LoongArch-toolchain-conventions-EN.html#_cc_preprocessor_built_in_macro_definitions)?

Thanks for clarifying. AFAICT, so far this is the only patch necessary for future binary compatibility between multiarch-aware and -unaware distributions (i.e. Debian & derivatives vs. the others). CPython itself builds and tests fine without this change; only the native extensions have different filename suffixes which is what this patch is trying to reconcile. Hence I don't think more `#ifdef`-ery should be necessary in the future.

Some more backgrounds though: LoongArch *might* need more treatment in the future if we'd want to provide perfect UX. There's two incompatible ABIs of LoongArch userland, so native libraries of different ABIs can't be interlinked without invoking UB, but Loongson wants the two ABIs (really *two worlds*) to share the same multiarch tuple. I expect some friction in the future if some user on one world uploads wheels to PyPI, only to be downloaded by someone on the other world, then failing at import time. It can be said to be a purely downstream problem though; most other upstream projects simply don't consider the "old world" / "ABI v1.0" to exist at all. And even if we do want to care, the changes here still apply: we've [already undergone debates](https://github.com/loongson/LoongArch-Documentation/pull/80) and this patch can be seen as the result. (This is kinda orthogonal to the issue at hand; I should probably post this elsewhere for further discussion.)

File: Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst
Comment (loongson-zn) [reply Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst]: Thanks, I will modify based on your suggestions.

**Comment by loongson-zn** [comment]
> @erlend-aasland cc @loongson-zn
> 
> But on the other hand, if this is the only patch that they will need, I am okay to accept the patch anyway.

@corona10  Currently, this is the only one patch. Because it affects binary compatibility, I believe it is necessary to merge as soon as possible.

**Comment by corona10** [comment]
> @corona10 Currently, this is the only one patch. Because it affects binary compatibility, I believe it is necessary to merge as soon as possible.

Yeah, let's pass the CI first, I am +1 with accepting this patch.

File: Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst
Comment (loongson-zn) [reply Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst]: @erlend-aasland After modifying *.rst, error occurred in test DOC CI test, `make V=1 -C Doc/ SPHINXOPTS="-q" SPHINXERRORHANDLING="-W --keep-going" html` need "do it twice", I have verified on PC.  What's the reason for CI error? What shoud I do?

**Comment by loongson-zn** [comment]
> Okay, let's pass the CI first, I am +1 with accepting this patch.
@corona10   Thank you for your support. I am debugging errors in CI Docs. I have asked  Erlend Aasland about the reason and wating for his reply. If you could tell me the reason, I would be very grateful

**Comment by loongson-zn** [comment]
This error disappeared after I executed it twice on my PC.

File: Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Build/2022-06-20-15-15-11.gh-issue-90656.kFBbKe.rst]: It seems Sphinx is picky about lists. I fixed the PR to use `*` instead of `-`. Let's see what the CI says now. I also copied the NEWS entry verbatim to What's New, and fixed the formatting (IMO, using monospaced font for the triplets is not needed; also, I used the architecture name from the Wiki page.)

**Review by erlend-aasland — APPROVED** [review]
Thanks, Zhang & WÁNG! I'll wait for Dong-hee's approval before landing this.

**Review by corona10 — APPROVED** [review]
lgtm

**Comment by erlend-aasland** [comment]
@xen0n, should we add your attributions to the NEWS and What's New entry also?

**Comment by xen0n** [comment]
> @xen0n, should we add your attributions to the NEWS and What's New entry also?

Either is OK, I'm unaffiliated with Loongson and in general just provides drive-by code reviews for everything LoongArch ;-)

In this case IMO let's see what @loongson-zn thinks though.

**Pull request merged**
